docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,125 @@
+- Special pods
+  - The Pod infrastructure container gets an OOM_SCORE_ADJ of -701
+  - Kubelet, Kube-Proxy, etc will get an even lower OOM_SCORE_ADJ
+
+Ensuring resource guarantees
+- After being configured suitably, the kernel will handle resource guarantees as specified in QoS Functionality
+
+Setting the OOM score
+- To set OOM_SCORE_ADJ of a container, we loop through all processes in the container, and set OOM_SCORE_ADJ
+- We keep looping until the list of processes in the container stabilizes
+
+## Issues and Assumptions
+
+- The QoS proposal assumes that swap memory is disabled. If swap is enabled, then resource guarantees (for pods that specify resource requirements) will not hold. For example, suppose 2 top-tier pods have reached their memory limit. They can start allocating memory on swap space. Eventually, if there isn’t enough swap space, processes in the pods might get killed. Ensure that swap space is disabled on our cluster setups.
+- We cannot set the OOM score of a container until it has launched. This could lead to races. For example, suppose that a top-tier pod is using 70% of the system’s memory, and another top-tier pod is using 30% of the system’s memory. A best-effort pod attempts to launch on the Kubelet. Initially the best-effort pod is using 2% of memory, and has an OOM_SCORE_ADJ of 0. So its OOM score is lower than the top-tier pod using 70% of system memory. The top-tier pod will be evicted by the best-effort pod. We think this isn’t a major issue, because it only comes up in a system that only has top-tier pods, and where a single top-tier pod uses over 50% of the system’s memory. The cleanest way to solve this problem is to add support for OOM score in cgroups, in the Linux kernel. This is out of scope.
+- We need to deal with the Kubelet’s restart policy when a process gets killed by the kernel. Suppose that we have 30 best-effort pods wanting to use 1GB of memory, with an “always” restart policy, in a system with 8GB of memory. With the current proposal, the system will keep launching best-effort pods, and the kernel will keep killing best-effort pods. In fact, the OOM killer is biased towards killing pods that are consuming more memory, so best-effort pods that are running would be evicted. In the first iteration, we will encourage users not to use the “always” restart policy with best-effort pods, but won’t handle the restart issue in Kubelet (Docker’s exponential backoff policy when restarting pods should mitigate the issue). Extension: In the final implementation, we will limit the number of times a pod can restart if it is OOM killed.
+- The OOM killer might kill processes across different pods. Suppose that 10 best-effort pods are running on the Kubelet, and the system runs out of memory. The OOM killer might kill one process from each pod. A better solution is to treat a pod as a unit - so that one of two entire pods are killed. This is out of scope, and difficult to solve reliably.

```

Comment:
I meant container, for the case where we'd actually restart it.

I agree that if we want an out-of-resource killer to cause a pod to move elsewhere, we'd need to kill the whole pod and report it as failed.